### PR TITLE
Completed Chapter 2

### PR DESCRIPTION
**Pull request by jlt10**
### Solutions for Chapter 2

I was wondering if there are situations where its preferred/cleaner to use the "go" pattern instead of pattern matching? Is it usually just a matter of preference/style?

cc @vrom911 @chshersh

### CODE REVIEW

File: src/Chapter2.hs [hunk 1]
```diff
@@ -605,7 +620,8 @@ Implement a function that duplicates each element of the list
 
 -}
 duplicate :: [a] -> [a]
-duplicate = error "duplicate: Not implemented!"
+duplicate [] = []
+duplicate (x:rest) = [x, x] ++ duplicate rest
```
Comment (chshersh) [inline src/Chapter2.hs]: It is more efficient to construct lists using the cons operator directly without introducing intermediate values.
```suggestion
duplicate (x:xs) = x : x : duplicate xs
```

File: src/Chapter2.hs [hunk 2]
```diff
@@ -620,7 +636,12 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven [x, _] = [x]
+takeEven (x:_:rest) = [x] ++ (takeEven rest)
```
Comment (chshersh) [inline src/Chapter2.hs]: The same comment about using `:` instead of `++` as for `duplicate` :slightly_smiling_face:

File: src/Chapter2.hs [hunk 3]
```diff
@@ -620,7 +636,12 @@ Write a function that takes elements of a list only on even positions.
 >>> takeEven [2, 1, 3, 5, 4]
 [2,3,4]
 -}
-takeEven = error "takeEven: Not implemented!"
+takeEven :: [a] -> [a]
+takeEven [] = []
+takeEven [x] = [x]
+takeEven [x, _] = [x]
```
Comment (chshersh) [inline src/Chapter2.hs]: `[x, _]` is syntax sugar for `(x : _ : [])` and you can see that this is already covered by your last line. It's a good idea to cover some small cases manually to see the general pattern first :slightly_smiling_face: But in the end, you can remove all redundant cases

File: src/Chapter2.hs [hunk 4]
```diff
@@ -727,7 +748,7 @@ value of the element itself
 🕯 HINT: Use combination of 'map' and 'replicate'
 -}
 smartReplicate :: [Int] -> [Int]
-smartReplicate l = error "smartReplicate: Not implemented!"
+smartReplicate l = concat (map (\x -> replicate x x) l)
```
Comment (chshersh) [inline src/Chapter2.hs]: Great! 👍🏼 
If you want to write this function even more elegantly, you can use the standard function `concatMap` which is a combination of `concat` and `map` as the name suggests 🙂
After that, you could notice, that you would be able to eta-reduce on the last argument :)

File: src/Chapter2.hs [hunk 5]
```diff
@@ -782,11 +803,12 @@ mastered the skill of eta-reducing.
...
 
--- TODO: type ;)
-listElementsLessThan x l = filter (< x) l
+listElementsLessThan :: Ord a => a -> [a] -> [a]
+listElementsLessThan x = filter (< x)
 
 -- Can you eta-reduce this one???
-pairMul xs ys = zipWith (*) xs ys
+pairMul :: Num a => [a] -> [a] -> [a]
+pairMul = zipWith (*)
```
Comment (chshersh) [inline src/Chapter2.hs:810]: Eta-reduction award! :trophy:

File: src/Chapter2.hs [hunk 6]
```diff
@@ -841,7 +863,10 @@ list.
 
 🕯 HINT: Use the 'cycle' function
 -}
-rotate = error "rotate: Not implemented!"
+rotate :: Int -> [a] -> [a]
+rotate n l 
+  | n < 0 = []
+  | otherwise = take (length l) (drop n (cycle l))
```
Comment (chshersh) [inline src/Chapter2.hs]: That's a correct solution! Though, it does a little bit more than work when the given number is big and the list is small. You will cycle through the list a lot. Maybe you can think of a way to optimize this :slightly_smiling_face:

File: src/Chapter2.hs [hunk 7]
```diff
@@ -857,7 +882,9 @@ and reverses it.
   function, but in this task, you need to implement it manually. No
   cheating!
 -}
-rewind = error "rewind: Not Implemented!"
+rewind :: [a] -> [a]
+rewind [] = []
+rewind (x: xs) = (rewind xs) ++ [x]
```
Comment (chshersh) [inline src/Chapter2.hs]: Your solution is correct! However, it is slow. In lists, it is quite slow to add anything at the end of the list. That is why it is always better to rewrite it with the `:` cons. Remember the explanation with trains? 🚂 🚋 🚋 

That is why a more efficient solution is with the accumulator and the recursive function that will do the addition at the start of the list which is instant!

You can read a bit more about the `go` pattern in here: https://kowainik.github.io/posts/haskell-mini-patterns#recursive-go

**Review by chshersh — APPROVED** [review]
Congrats on finishing Chapter Two!
And well done with the advanced task :clap: 

I left my review and some comments and suggestions :slightly_smiling_face: 

Answering your question about `go` vs pattern-matching. It's not actually a `vs` because eventually recursive `go` functions use pattern matching. The "Recursive go" pattern shines when you need to pass some additional parameters to recursive calls (maintain some state between recursive calls). Or when you need to perform some transformations on the initial data only once. E.g. the `firstDigit` function from Chapter One is a perfect candidate for the "Recursive go" pattern because you can handle all corner cases in the beginning and then just leave the work to the recursive call.

**Comment by jlt10** [comment]
@chshersh  Thanks for taking the time to answer my questions and leave detailed comments! This is my first time learning a functional language, so I really appreciate the performance notes :D

**Pull request merged**